examples/raycasting/main.go

```diff
@@ -0,0 +1,312 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build example jsgo
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"image"
+	"image/color"
+	_ "image/png"
+	"log"
+	"math"
+	"sort"
+
+	"github.com/hajimehoshi/ebiten"
+	"github.com/hajimehoshi/ebiten/ebitenutil"
+	"github.com/hajimehoshi/ebiten/examples/resources/images"
+	"github.com/hajimehoshi/ebiten/inpututil"
+)
+
+const (
+	screenWidth  = 240
+	screenHeight = 240
+)
+
+var (
+	bgImage       *ebiten.Image
+	shadowImage   *ebiten.Image
+	triangleImage *ebiten.Image
+)
+
+func init() {
+	// Decode image from a byte slice instead of a file so that
+	// this example works in any working directory.
+	// If you want to use a file, there are some options:
+	// 1) Use os.Open and pass the file to the image decoder.
+	//    This is a very regular way, but doesn't work on browsers.
+	// 2) Use ebitenutil.OpenFile and pass the file to the image decoder.
+	//    This works even on browsers.
+	// 3) Use ebitenutil.NewImageFromFile to create an ebiten.Image directly from a file.
+	//    This also works on browsers.
+	img, _, err := image.Decode(bytes.NewReader(images.Tile_png))
+	if err != nil {
+		log.Fatal(err)
+	}
+	bgImage, _ = ebiten.NewImageFromImage(img, ebiten.FilterDefault)
+	shadowImage, _ = ebiten.NewImage(screenWidth, screenHeight, ebiten.FilterDefault)
+	triangleImage, _ = ebiten.NewImage(screenWidth, screenHeight, ebiten.FilterDefault)
+	triangleImage.Fill(color.White)
+}
+
+type Line struct {
+	X1, Y1, X2, Y2 float64
+}
+
+func newRay(x, y, length, angle float64) Line {
+	return Line{
+		X1: x,
+		Y1: y,
+		X2: x + length*math.Cos(angle),
+		Y2: y + length*math.Sin(angle),
+	}
+}
+
+// intersection calculates the intersection of given two lines.
+// https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection#Given_two_points_on_each_line
+func intersection(l1, l2 Line) (float64, float64, error) {
+	denom := (l1.X1-l1.X2)*(l2.Y1-l2.Y2) - (l1.Y1-l1.Y2)*(l2.X1-l2.X2)
+	tNum := (l1.X1-l2.X1)*(l2.Y1-l2.Y2) - (l1.Y1-l2.Y1)*(l2.X1-l2.X2)
+	uNum := -((l1.X1-l1.X2)*(l1.Y1-l2.Y1) - (l1.Y1-l1.Y2)*(l1.X1-l2.X1))
+
+	if denom == 0 {
+		return 0, 0, errors.New("lines parallel or coincident")
+	}
+
+	t := tNum / denom
+	if t > 1 || t < 0 {
+		return 0, 0, errors.New("lines intersect, segments do not")
+	}
+
+	u := uNum / denom
+	if u > 1 || u < 0 {
+		return 0, 0, errors.New("lines intersect, segments do not")
+	}
+
+	x := l1.X1 + t*(l1.X2-l1.X1)
+	y := l1.Y1 + t*(l1.Y2-l1.Y1)
+	return x, y, nil
+}
+
+func calcAngle(l Line) float64 {
+	return math.Atan2(l.Y2-l.Y1, l.X2-l.X1)
+}
+
+// rayCasting returns a slice of Line originating from point cx, cy and intersecting with objects
+func rayCasting(cx, cy float64, objects [][]Line) []Line {
+	var rays []Line
+
+	rayLength := 1000 // something large
```

Comment:
I meant `const rayLength = 1000`

Reply:
Ah! Should I keep it inside the function?

Reply:
made the comment a bit more clear (I think)

Reply:
That's right! (`rayLength := 1000` cannot be compiled due to type mismatch anyway)